File: GeoJSON4EntityFramework/Elements/Polygon.vb
#If EF5 Then
Imports System.Data.Spatial
#End If

#If EF6 Then
Imports System.Data.Entity.Spatial
#End If

Public Class Polygon
    Inherits GeoJsonGeometry

    <Newtonsoft.Json.JsonIgnore()>
    Public Property Rings As New List(Of CoordinateList)

    Public Overrides ReadOnly Property Coordinates()
        Get
            Return Rings
        End Get
    End Property

    Public Overrides Function Transform(xform As CoordinateTransform) As GeoJsonGeometry
        If xform Is Nothing Then
            Throw New ArgumentNullException(NameOf(xform))
        End If

        Dim poly As New Polygon()
        If Not Rings Is Nothing Then
            poly.Rings.AddRange(Rings.Select(Function(ring) ring.CloneList(xform)))
        End If

        If Not BoundingBox Is Nothing Then
            poly.BoundingBox = TransformFunctions.TransformBoundingBox(BoundingBox, xform)
        End If
        Return poly
    End Function

    Private Function RingToCoordinateList(ring As DbGeometry) As CoordinateList
        Dim extRingCoords As New CoordinateList()
        For i = 1 To ring.PointCount
            Dim pt = ring.PointAt(i)
            extRingCoords.Add(New Coordinate(pt.XCoordinate, pt.YCoordinate))
        Next
        Return extRingCoords
    End Function

    Public Overrides Sub CreateFromDbGeometry(inp As DbGeometry)
        If inp.SpatialTypeName <> TypeName Then Throw New ArgumentException
        Rings.Clear()

        ' Process exterior ring
        Dim extRing = inp.ExteriorRing
        Rings.Add(RingToCoordinateList(extRing))

        ' Process interior rings (ie. holes)
        If inp.InteriorRingCount > 0 Then
            For i = 1 To inp.InteriorRingCount
                Dim intRing = inp.InteriorRingAt(i)
                Rings.Add(RingToCoordinateList(intRing))
            Next
        End If
    End Sub
End Class

File: GeoJSON4EntityFrameworkCS/Elements/Polygon.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using GeoJSON4EntityFramework.Base;

namespace GeoJSON4EntityFramework.Elements
{
    public class Polygon : GeoJsonGeometry<Polygon>, IGeoJsonGeometry
    {

        [Newtonsoft.Json.JsonIgnore]
        public List<CoordinateList> PointsList { get; set; }

        public override object Coordinates
        {
            get {
			try {
				if (PointsList.Count == 0) {
					return new double[0];
				} else {
					var @out = new double[PointsList.Count][][];
					for (var k = 0; k <= PointsList.Count - 1; k++)
					{
					    var points3 = PointsList[k];
					    switch (points3.Count)
					    {
					        case 0:
					            return new double[0];
					        case 1:
					            throw new Exception("There must be an array of two or more points");
					        default:
					            @out[k] = new double[points3.Count][];
					            Parallel.For(0, points3.Count, i => { @out[k][i] = points3[i].Coordinates; });
					            break;
					    }
					}
				    return @out;
				}
			} catch (Exception) {
				return new double[0];
			}
		}
        }

        public override void CreateFromDbGeometry(DbGeometryWrapper inp)
        {
            if (inp.Geometry.SpatialTypeName != TypeName)
                throw new ArgumentException();
            PointsList.Clear();

            Ring2Coordinates(new DbGeometryWrapper(inp.Geometry.ExteriorRing));
            dynamic numRings = inp.Geometry.InteriorRingCount;
            if ((numRings == 0))
                return;
            for (var i = 1; i <= numRings; i++)
            {
                dynamic ring = inp.Geometry.InteriorRingAt(i);
                Ring2Coordinates(new DbGeometryWrapper(ring));
            }

        }

        private void Ring2Coordinates(DbGeometryWrapper ring)
        {
            dynamic points = new CoordinateList();
            for (var i = 1; i <= ring.Geometry.PointCount; i++)
            {
                dynamic point = ring.Geometry.PointAt(i);
                points.AddNew(point.XCoordinate, point.YCoordinate);
            }
            PointsList.Add(points);
        }

    }

}
